packages/hint-compat-api/src/css-base.ts

```diff
@@ -0,0 +1,160 @@
+import { CompatApi, CompatCSS } from './helpers';
+import { FeatureInfo, BrowsersInfo, SupportStatementResult } from './types';
+import { SimpleSupportStatement, VersionValue, SupportBlock, SupportStatement } from './types-mdn.temp';
+
+import { browserVersions } from './helpers/normalize-version';
+import { CSSFeatureStatus } from './enums';
+
+export default abstract class BaseCompatApiCSS implements IHint {
+    private readonly statusName: CSSFeatureStatus;
+    private compatApi: CompatApi;
+    private compatCSS: CompatCSS;
+
+    abstract getFeatureVersionValueToAnalyze(browserFeatureSupported: SimpleSupportStatement): VersionValue;
+    abstract isSupportedVersion(browser: BrowsersInfo, feature: FeatureInfo, currentVersion: number, version: number): boolean;
+    abstract isVersionValueSupported(version: VersionValue): boolean;
+    abstract isVersionValueTestable(version: VersionValue): boolean;
+
+    public constructor(context: HintContext<StyleEvents>, statusName: CSSFeatureStatus, isCheckingNotBroadlySupported: boolean) {
+        this.compatApi = new CompatApi('css', context, isCheckingNotBroadlySupported);
+        this.compatCSS = new CompatCSS(context, this.testFeatureIsSupported.bind(this));
+        this.statusName = statusName;
+
+        context.on('parse::end::css', this.onParseCSS.bind(this));
+    }
+
+    private async onParseCSS(styleParse: StyleParse): Promise<void> {
+        const { resource } = styleParse;
+
+        this.compatCSS.setResource(resource);
+        await this.compatCSS.searchCSSFeatures(this.compatApi.compatDataApi, styleParse);
+    }
+
+    private async testFeatureIsSupported(feature: FeatureInfo, supportBlock: SupportBlock): Promise<void> {
+        const browsersToSupport = Object.entries(supportBlock).filter(([browserName, _]: [string, SupportStatement]): boolean => {
+            return this.compatApi.isBrowserIncludedInCollection(browserName);
+        });
+
+        const groupedSupportByBrowser = browsersToSupport.reduce((group, browserInfo, _, a1) => {
+            return this.groupSupportStatementByBrowser(feature, group, browserInfo);
+        }, {});
+
+        const supportStatementResult: SupportStatementResult = {
+            browsersToSupportCount: browsersToSupport.length,
+            groupedBrowserSupport: groupedSupportByBrowser,
+            notSupportedBrowsersCount: Object.keys(groupedSupportByBrowser).length
+        };
+
+        const hasImcompatibleBrowsers: boolean = supportStatementResult.notSupportedBrowsersCount > 0;
+
+        if (!hasImcompatibleBrowsers) {
+            return;
+        }
+
+        const message: string = this.generateReportErrorMessage(feature, supportStatementResult);
+
+        await this.compatCSS.reportError(feature, message);
+    }
+
+    private groupSupportStatementByBrowser(feature: FeatureInfo, group: { [browserName: string]: string[] }, browserInfo: [string, SupportStatement]) {
```

Comment:
browserInfor make me confused, because have the same noun as type BrowserInfo, but is not the same, probably we can find an other name. And other point, it isn't  recommendable use the words 'Info','Data', etc in variables, it can make a lot of confusions, because we don't know what kind of things contains the object.

Well, we can discus about it.

Reply:
Yes, you are right. All these words are a bit ambiguous